client/github_client_manager.go

```diff
@@ -0,0 +1,115 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/bradleyfalzon/ghinstallation/v2"
+	"github.com/google/go-github/v45/github"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/mattermost/release-bot/version"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	ClientCacheSize = 64
+)
+
+type GithubClientManager interface {
+	Get(installationID int64) (*github.Client, error)
+	CreateToken(installationID int64) (AccessToken, error)
+}
+
+type AccessToken interface {
+	IsExpired() bool
+	GetToken() string
+}
+
+type clientCache struct {
+	appID              int64
+	privateKeyFile     string
+	userAgent          string
+	cache              *lru.Cache
+	transport          http.RoundTripper
+	appClient          *github.Client
+	installationTokens map[int64]AccessToken
+}
+
+type accessToken struct {
+	token     string
+	expiresAt time.Time
+}
+
+func NewManager(appID int64, privateKeyFile string) (GithubClientManager, error) {
+	cache, err := lru.New(ClientCacheSize)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create cache")
+	}
+	version := version.Full()
+	itr, err := ghinstallation.NewAppsTransportKeyFromFile(http.DefaultTransport, appID, privateKeyFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Can not initialize GitHub client! Check configuration values.")
+	}
+
+	return &clientCache{
+		appID:              appID,
+		privateKeyFile:     privateKeyFile,
+		cache:              cache,
+		transport:          http.DefaultTransport,
+		userAgent:          fmt.Sprintf("%s/%s", version.Name, version.Version),
+		appClient:          github.NewClient(&http.Client{Transport: itr}),
+		installationTokens: make(map[int64]AccessToken),
+	}, nil
+}
+
+func (cc *clientCache) Get(installationID int64) (*github.Client, error) {
+	cli, ok := cc.cache.Get(installationID)
+	if ok {
+		return cli.(*github.Client), nil
+	}
+	itr, err := ghinstallation.NewKeyFromFile(
+		cc.transport,
+		cc.appID,
+		installationID,
+		cc.privateKeyFile,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Can not initialize GitHub client! Check configuration values.")
+	}
+	client := github.NewClient(&http.Client{Transport: itr})
+	client.UserAgent = cc.userAgent
+	cc.cache.Add(installationID, client)
+	return client, nil
+}
+
+func (cc *clientCache) CreateToken(installationID int64) (AccessToken, error) {
+	token, found := cc.installationTokens[installationID]
+	if found && !token.IsExpired() {
+		log.Info("Using non-expired repository github token")
+		return token, nil
+	}
+	ghToken, _, err := cc.appClient.Apps.CreateInstallationToken(context.Background(), installationID, &github.InstallationTokenOptions{})
```

Comment:
Since `InstallationTokenOptions` is empty, this token will have access to the entire installation, right? So are we scoping the permissions by having separate installations for each repo, or should we actually specify the repositories here?

Reply:
Yes the token can access to entire installation. And we should create dedicated installation per repository. In scope of this, no need to define options,however we can specify repositories to make it visible here.

Installation is mentioned at `GitHub Application Installation, Permissions and Hooks
` section at our document. `Github Application will be configured at our Enterprise Account and enabled per repository to have dedicated installation identifiers per repository.`